### DISCUSSION TITLE
Potential issue with "tethys manage collectstatic"

### DISCUSSION THREAD

**Discussion by shawncrawley** [tethysplatform/tethys#636] [Q&A]
I installed an application into a production environment per the [official documentation](http://docs.tethysplatform.org/en/stable/installation/production/app_installation.html). Maybe I'm not using the app_workspace folder as intended, but I have a file stored in my app_workspace directory that gets frequently modified as part of our development lifecycle. When trying to deploy changes to the application, I have been running:
``` bash
git pull
tethys install
tethys manage collectall -f
```
First issue is that passing the "-f" option still prompts me as to whether I am sure or not and requires me to input "yes" or "no". Another issue - the bigger one - is that the changes to the file in app_workspaces do not get reflected in the <TETHYS_WORKSPACES_ROOT> directory. I have been having to manually copy the file over.

Am I misunderstanding how this is all supposed to work?

**Comment by swainn** [comment]
Hey @shawncrawley. Regarding the `-f` option issue I think what you want is the `--noinput` option. It does seem odd for a `--force` option to prompt for input. It's also not extremely clear what the ``--noinput`` option does.

I don't understand fully what is going on to answer the second question, so here are a few questions:

1. What OS are you running in the production environment? Symbolic links won't work on Windows.
2. How are you using workspaces? Are you using the dedicated workspaces API (i.e.: the app_workspaces decorator) or are you hard-coding relative paths?
3. Does the app seem to have access to the updated file before you copy it to TETHYS_WORKSPACES_ROOT? after?
4. Are any files generated/changed in the workspace at runtime? Or is it only when you update the app in development that changes happen?
5. If you generating/changing files at runtime, do they appear in the TETHYS_WORKSPACES_ROOT location?

**Reply by shawncrawley** [reply]
@swainn,

First, there's definitely some ambiguity surrounding both the `--noinput` and the `--force` options. But yeah, I think in this case I actually wanted `--force`.

To answer your questions:
1. The production installation is on RHEL 7
2. I'm using the dedicated Workspaces API by way of the decorator. More specifically, I am storing a version-controlled config file in the app_workspaces directory. I chose to store it there because my home controller needs to check settings in this file in order to properly render home.hmtl.
3. It seems that upon initial installation of the application - following the initial execution of `tethys manage collectall`, the file ended up where it was supposed to be (i.e. in the TETHYS_WORKSPACES_ROOT directory) and the application could find it. It was only when pulling changes to the repository that included changes to this file and running `tethys manage collectall` again that I found that the file in TETHYS_WORKSPACES_ROOT did not reflect the changes (i.e. it was still the old version).
4. I do not use the workspaces to store dynamically-generated files at runtime. There should only ever be the single config file in the app_workspaces directory, and it should only ever change through `git pull`.
5. This question is irrelevant due to my number 4 answer above.

**Comment by swainn** [comment]
@shawncrawley after looking at the code (since the ambiguous CLI help was no use) I discovered `collectworkspaces` will behave as follows:

* If you run `tethys manage collectworkspaces` (without `--force`) AND the workspaces already exists in `TETHYS_WORKSPACES_ROOT`:
     * the copy of the workspace in your app's install directory (with your updated changes) is deleted and
     * the link back to the `TETHYS_WORKSPACES_ROOT` is re-created in its place.

* If you run `tethys manage collectworkspaces --force` AND the workspaces already exists in `TETHYS_WORKSPACES_ROOT`:
    * the copy of the workspace in `TETHYS_WORKSPACE_ROOT` is deleted,
    * the workspace in your app's install directory is copied to `TETHYS_WORKSPACE_ROOT`, and
    * the workspace your app's directory is then removed and the link back to the `TETHYS_WORKSPACES_ROOT` is created in its place.

The effect is that your new changes seem to be ignored if you use `tethys manage collectworkspaces` or you lose all data generated in your workspaces by the app at runtime and get only your new files if you use `tethys manage collectworkspaces --force`.

There is no way to "sync" new changes into an existing collected workspace at the moment.

**Reply by swainn** [reply]
Here are a few additional thoughts on this:

1. When you run `tethys install` to install the new version of your app, the link to `TETHYS_WORKSPACES_ROOT` overwritten by the workspace directory (75% confidence).
2. As I've thought about this more, I think it was designed to work this way. The original conception was that workspaces were for files generated at runtime and thus store some "state" of the app, like a database. We didn't plan for it to be used as a configuration store with updates being synced into it periodically.

So, technically, you are misusing workspaces for storing this kind of file. Whether that should be the case or not is up for discussion, which is why I opened issue #645.

Here are a few suggestions for these types of files:
1. Create a `resource` or `config` folder (don't forget to add it as a resource directory in your setup.py)
2. Use App Settings for storing configuration parameters
